physical/redis.go

```diff
@@ -0,0 +1,370 @@
+package physical
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/garyburd/redigo/redis"
+)
+
+// RedisBackend is a physical backend that stores data at a specific prefix
+// within Redis.  This is not especially useful as an ordinary
+// secret-storage backend, unless you set up Redis persistence.
+//
+// But this backend offers a relatively simple way to set up a "medium
+// availability" Vault cluster, where you use S3 or another backend for
+// storage, and a single Redis server as a "ha_backend" to handle leader
+// elections.  If the Redis server fails, then all vault servers should go
+// to standby until Redis returns.
+//
+// This is potentially useful because Redis servers are ubiquitous and
+// surprisingly reliable, and because not everybody wants to set up and
+// administer either Consul or etcd for a single leader lock.  Also,
+// high-quality hosted and managed Redis services are widely available.
+type RedisBackend struct {
+	path     string
+	pool     *redis.Pool
+	logger   *log.Logger
+	lockConf *RedisLockConfig
+}
+
+// Configuration options for a RedisLock.
+type RedisLockConfig struct {
+	// How long should the leader TTL last?
+	leaderTTL               time.Duration
+	// How long should the leader wait before renewing the TTL?  This
+	// should be no more than half of leaderTTL.
+	leaderTTLRenewInterval  time.Duration
+	// How often should a standby server attempt to take the lock?
+	leaderLockRetryInterval time.Duration
+}
+
+func parseRedisTimeout(conf map[string]string, key string, defaultVal time.Duration) (time.Duration, error) {
+	timeoutStr, ok := conf[key]
+	if !ok {
+		return defaultVal, nil
+	}
+	timeout, err := strconv.Atoi(timeoutStr)
+	if err != nil {
+		return 0, err
+	}
+	return time.Duration(timeout) * time.Millisecond, nil
+}
+
+func newRedisBackend(conf map[string]string, logger *log.Logger) (Backend, error) {
+	// Get the Redis path from the configuration.
+	path, ok := conf["path"]
+	if !ok {
+		path = "vault"
+	}
+
+	// Get the Redis URL from either our config file or our environment.
+	url, ok := conf["url"]
+	if !ok {
+		url = "redis://127.0.0.1:6379"
+	}
+	urlEnv := os.Getenv("REDIS_URL")
+	if urlEnv != "" {
+		url = urlEnv
+	}
+
+	// Get our lock timeouts.
+	leaderTTL, err := parseRedisTimeout(conf, "leader_ttl", 30*time.Second)
+	if err != nil {
+		return nil, err
+	}
+	leaderTTLRenewInterval, err := parseRedisTimeout(conf, "leader_ttl_renew_interval", 1*time.Second)
+	if err != nil {
+		return nil, err
+	}
+	leaderLockRetryInterval, err := parseRedisTimeout(conf, "leader_lock_retry_interval", 5*time.Second)
+	if err != nil {
+		return nil, err
+	}
+	if 2*leaderTTLRenewInterval > leaderTTL {
+		return nil, fmt.Errorf("leader_ttl_renew_interval must be no more than half of leader_ttl")
```

Comment:
Yes, to help ensure correctness. The leader only notices a dead lock when renewing the TTL, and it will need some margin of time to step down. So `leaderTTLRenewInterval == leaderTTL` raises the risks of corruption by an inattentive former leader, and `leaderTTLRenewInterval > leaderTTL` guarantees invalidity.

Also, a max time of `ttl/2.0` for renewals is a very common vault suggestion.